src/AspNet.Security.OAuth.Deezer/DeezerAuthenticationHandler.cs

```diff
@@ -0,0 +1,154 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Deezer
+{
+    public class DeezerAuthenticationHandler : OAuthHandler<DeezerAuthenticationOptions>
+    {
+        public DeezerAuthenticationHandler(
+            [NotNull] IOptionsMonitor<DeezerAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        private static readonly RandomNumberGenerator CryptoRandom = RandomNumberGenerator.Create();
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(OAuthCodeExchangeContext context)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "app_id", Options.App_Id },
+                { "secret", Options.Secret },
+                { "code", context.Code },
+                { "output", "json" }
+            };
+
+            // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+            if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out string codeVerifier))
+            {
+                tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier);
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            string endpoint = QueryHelpers.AddQueryString(Options.TokenEndpoint, tokenRequestParameters);
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Get, endpoint);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode)
+            {
+                var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+                return OAuthTokenResponse.Success(payload);
+            }
+            else
+            {
+                string error = "OAuth token endpoint failure: " + await Display(response);
```

Comment:
Please use the same error pattern as on line 82.

Reply:
Is it ok so? 
```
if (!response.IsSuccessStatusCode)
    {
        const string error = "An error occurred while retrieving an OAuth token";

        Logger.LogError("{Error}: the remote server " +
                        "returned a {Status} response with the following payload: {Headers} {Body}.",
                        /* Error: */  error,
                        /* Status: */ response.StatusCode,
                        /* Headers: */ response.Headers.ToString(),
                        /* Body: */ await response.Content.ReadAsStringAsync());

        return OAuthTokenResponse.Failed(new Exception(error));
    }
```

Or should I throw better a HttpRequestException?

Reply:
That's fine - it matches what all the other calls in the other providers do.